**Лабораторная работа №1**

In [1]:
import numpy as np # Массивы (матрицы, векторы, линейная алгебра)
import matplotlib.pyplot as plt # Научная графика
%matplotlib inline
    # Говорим jupyter'у, чтобы весь графический вывод был в браузере, а не в отдельном окне
import pandas as pd             # Таблицы и временные ряды (dataframe, series)
import seaborn as sns           # Еще больше красивой графики для визуализации данных
import sklearn                  # Алгоритмы машинного обучения
plt.style.use('default')

**Описание задачи:**

 Данными для задачи является набор объявлений о продаже 51783  автомобилей Ford на территории США в районе Айовы. Данные собранны в декабре 2018 года и  взяты с сайиа Cars.com.
 
 Постановка задачи:
 
 С помощью методов машинного обучения, научить модель определять drivetrain - тип трансмиссии.
 
  
**Типы трансмисии:**

4WD - 

FWD -

RWD -

AWD -

Four Wheel Drive -



**Загружаем данные:**

In [2]:
url = "https://raw.githubusercontent.com/SlavikLenkin/MachineLearning/main/carData.csv"
full_data = pd.read_csv(url)# Читаем данные

In [3]:
full_data.shape 

(51783, 13)

Отображаем первые строки таблицы (по умолчанию 5):

In [4]:
full_data.head()

price         miles fuel_type      exterior_color  \
0  $25,290       6 miles    Hybrid        Oxford White   
1  $12,988  51,713 miles    Hybrid                Gray   
2  $25,995      10 miles    Hybrid               White   
3  $20,225   2,022 miles    Hybrid   Magnetic Metallic   
4   $9,413  48,164 miles    Hybrid        Oxford White   

                          interior_color drivetrain    transmission  \
0                         Charcoal Black        FWD             CVT   
1                         Charcoal Black        FWD   Automatic CVT   
2                                  Black        FWD   Automatic CVT   
3                         Charcoal Black        FWD   Automatic CVT   
4   Medium Light Stone/Medium Dark Stone        FWD       AUTOMATIC   

                          engine                 VIN  \
0                     4 Cylinder   1FADP5AU6JL100444   
1   2.0L I4 16V MPFI DOHC Hybrid   1FADP5BUXDL528553   
2   2.0L I4 16V MPFI DOHC Hybrid   1FADP5BU7GL115512   
3   2.0L I4 16V MPFI DOHC Hybrid   1FADP5AUXJL105758   
4         4 Cylinder Engine 2.0L   1FADP5AU3DL540626   

                              name                              sellerAddress  \
0    New 2018 Ford C-Max Hybrid SE  2406 N Jefferson Way, Indianola, IA 50125   
1  Used 2013 Ford C-Max Hybrid SEL         2020 W 20th St S, Newton, IA 50208   
2   New 2016 Ford C-Max Hybrid SEL             916 W St S, Grinnell, IA 50112   
3    New 2018 Ford C-Max Hybrid SE        600 Brooks Rd, Iowa Falls, IA 50126   
4   Used 2013 Ford C-Max Hybrid SE       215 15th St SW, Mason City, IA 50401   

          id     modelName  
0  719242528  C-Max Hybrid  
1  753737631  C-Max Hybrid  
2  675439230  C-Max Hybrid  
3  738620869  C-Max Hybrid  
4  754880903  C-Max Hybrid

Отображаем последние строки таблицы (по умолчанию 5):

In [5]:
full_data.tail()

price          miles       fuel_type          exterior_color  \
51778  $54,793            NaN          Diesel            Shadow Black   
51779  $64,711            NaN          Diesel                Magnetic   
51780  $19,545   50,376 miles   E85 Flex Fuel            Oxford White   
51781  $58,987   19,189 miles          Diesel   Ingot Silver Metallic   
51782  $22,987  106,446 miles   E85 Flex Fuel            Oxford White   

           interior_color drivetrain                  transmission  \
51778   Medium Earth Gray        4WD             6-Speed Automatic   
51779               Black        4WD             6-Speed Automatic   
51780               Steel        RWD   TorqShift 6-Speed Automatic   
51781               Black        4WD   TorqShift 6-Speed Automatic   
51782               Steel    4 WHEEL                     AUTOMATIC   

                                             engine                 VIN  \
51778              6.7L V8 32V DDI OHV Turbo Diesel   1FT7W2BT9JEC40281   
51779              6.7L V8 32V DDI OHV Turbo Diesel   1FT7W2BT5JEC07424   
51780                6.2L V8 EFI SOHC 16V Flex Fuel   1FDBF2A65DEB65326   
51781   Power Stroke 6.7L V8 DI 32V OHV Turbodiesel   1FT7W2BT9HEF37240   
51782                         6.2L 8 CYL. FLEX FUEL   1FT7W2B69DEA17804   

                              name  \
51778      New 2018 Ford F-250 XLT   
51779   New 2018 Ford F-250 Lariat   
51780      Used 2013 Ford F-250 XL   
51781  Used 2017 Ford F-250 Lariat   
51782     Used 2013 Ford F-250 XLT   

                                    sellerAddress         id modelName  
51778    2860 S Noland Rd, Independence, MO 64055  751165085     F-250  
51779    2860 S Noland Rd, Independence, MO 64055  737236411     F-250  
51780       2245 SW Topeka Blvd, Topeka, KS 66611  749747000     F-250  
51781       2245 SW Topeka Blvd, Topeka, KS 66611  756269741     F-250  
51782  1500 NW S Outer Rd, Blue Springs, MO 64015  756360809     F-250

**Обзор данных и их типов:**
1. price - цена автомобиля,количественный
2. miles - пробег,количественный
3. fuel_type - тип топлива,номинальный
4. exterior_color - цвет кузова,номинальный
5. interior_color - цвет салона,номинальный
6. drivetrain - тип трансмиссии,номинальный
7. transmission - тип коробки передач,номинальный
8. engine - тип дивигателя,номинальный
9. VIN - уникальный идентификационный номер машины,номинальный
10. name - названи машины,номинальный
11. sellerAddress - адресс продавца,номинальный
12. id - уникальный номер объявления,номинальный
13. modelName - название модели,номинальный



In [6]:
full_data.columns

Index(['price', 'miles', 'fuel_type', 'exterior_color', 'interior_color',
       'drivetrain', 'transmission', 'engine', 'VIN', 'name', 'sellerAddress',
       'id', 'modelName'],
      dtype='object')

**Диапазоны индексов:**

In [7]:
full_data.index

RangeIndex(start=0, stop=51783, step=1)

**Достпук к отдельным элементам и выделение подтаблиц(slice):**

In [8]:
full_data.at[3, 'fuel_type']

' Hybrid'

In [9]:
full_data.loc[3, 'fuel_type']

' Hybrid'

In [10]:
full_data.iat[3, 2]

' Hybrid'

In [11]:
full_data.iloc[3, 2]

' Hybrid'

**Подтаблицы:**

In [12]:
full_data.loc[[13,123,414,2222],['fuel_type','engine','transmission']]

fuel_type                           engine  \
13            Hybrid                        I4 Hybrid   
123         Gasoline                          4.0L V6   
414         Gasoline   Regular Unleaded V-8 5.0 L/302   
2222   E85 Flex Fuel                      2.5L 4 cyls   

                  transmission  
13                         CVT  
123          Automatic 4-Speed  
414    10-Speed Automatic w/OD  
2222         Automatic 6-Speed

In [13]:
full_data.iloc[[13,123,414,2222],[2,6,7]]

fuel_type              transmission  \
13            Hybrid                       CVT   
123         Gasoline         Automatic 4-Speed   
414         Gasoline   10-Speed Automatic w/OD   
2222   E85 Flex Fuel         Automatic 6-Speed   

                               engine  
13                          I4 Hybrid  
123                           4.0L V6  
414    Regular Unleaded V-8 5.0 L/302  
2222                      2.5L 4 cyls

In [14]:
full_data.loc[[13,123,414,2222],['fuel_type','engine','transmission']].loc[123,'transmission']

' Automatic 4-Speed'

In [15]:
full_data.iloc[[13,123,414,2222],[2,6,7]].iloc[1,1]

' Automatic 4-Speed'

**Целые строки и столбцы:**

In [16]:
full_data.loc[:,['fuel_type','engine','transmission']]

fuel_type                                        engine  \
0              Hybrid                                    4 Cylinder   
1              Hybrid                  2.0L I4 16V MPFI DOHC Hybrid   
2              Hybrid                  2.0L I4 16V MPFI DOHC Hybrid   
3              Hybrid                  2.0L I4 16V MPFI DOHC Hybrid   
4              Hybrid                        4 Cylinder Engine 2.0L   
...               ...                                           ...   
51778          Diesel              6.7L V8 32V DDI OHV Turbo Diesel   
51779          Diesel              6.7L V8 32V DDI OHV Turbo Diesel   
51780   E85 Flex Fuel                6.2L V8 EFI SOHC 16V Flex Fuel   
51781          Diesel   Power Stroke 6.7L V8 DI 32V OHV Turbodiesel   
51782   E85 Flex Fuel                         6.2L 8 CYL. FLEX FUEL   

                       transmission  
0                               CVT  
1                     Automatic CVT  
2                     Automatic CVT  
3                     Automatic CVT  
4                         AUTOMATIC  
...                             ...  
51778             6-Speed Automatic  
51779             6-Speed Automatic  
51780   TorqShift 6-Speed Automatic  
51781   TorqShift 6-Speed Automatic  
51782                     AUTOMATIC  

[51783 rows x 3 columns]

In [17]:
full_data.loc[[12,10000],:]

price         miles  fuel_type                     exterior_color  \
12     $12,494  53,513 miles     Hybrid                              White   
10000  $28,245       5 miles   Gasoline   White Platinum Tri-Coat Metallic   

            interior_color drivetrain        transmission  \
12      Medium Light Stone        FWD       Automatic CVT   
10000                Ebony        FWD   6-Speed Automatic   

                              engine                 VIN  \
12      2.0L I4 16V MPFI DOHC Hybrid   1FADP5AU0FL123175   
10000     1.5L I4 16V GDI DOHC Turbo   3FA6P0CDXKR133426   

                                 name  \
12     Used 2015 Ford C-Max Hybrid SE   
10000        New 2019 Ford Fusion SEL   

                                  sellerAddress         id     modelName  
12            6345 N 28th St, Lincoln, NE 68504  750700942  C-Max Hybrid  
10000  3200 S. Outer Rd, Blue Springs, MO 64015  753582800        Fusion

In [18]:
full_data['fuel_type']

0                Hybrid
1                Hybrid
2                Hybrid
3                Hybrid
4                Hybrid
              ...      
51778            Diesel
51779            Diesel
51780     E85 Flex Fuel
51781            Diesel
51782     E85 Flex Fuel
Name: fuel_type, Length: 51783, dtype: object

In [19]:
full_data[['fuel_type','engine']]

fuel_type                                        engine
0              Hybrid                                    4 Cylinder
1              Hybrid                  2.0L I4 16V MPFI DOHC Hybrid
2              Hybrid                  2.0L I4 16V MPFI DOHC Hybrid
3              Hybrid                  2.0L I4 16V MPFI DOHC Hybrid
4              Hybrid                        4 Cylinder Engine 2.0L
...               ...                                           ...
51778          Diesel              6.7L V8 32V DDI OHV Turbo Diesel
51779          Diesel              6.7L V8 32V DDI OHV Turbo Diesel
51780   E85 Flex Fuel                6.2L V8 EFI SOHC 16V Flex Fuel
51781          Diesel   Power Stroke 6.7L V8 DI 32V OHV Turbodiesel
51782   E85 Flex Fuel                         6.2L 8 CYL. FLEX FUEL

[51783 rows x 2 columns]

**Доступ по булевой майске**

In [20]:
df= full_data.iloc[4702:4709,:4]
df

price          miles                fuel_type  \
4702  $23,395   10,393 miles                 Gasoline   
4703  $54,647       10 miles                   Diesel   
4704   $8,488  213,601 miles                  Unknown   
4705   $2,500  211,853 miles                 Gasoline   
4706   $9,950    6,165 miles                   Diesel   
4707  $29,869        4 miles   Compressed Natural Gas   
4708  $17,288   86,729 miles                 Gasoline   

                          exterior_color  
4702                        Summit White  
4703   Brilliant Black Crystal Pearlcoat  
4704                               White  
4705                                Blue  
4706                                 Red  
4707                        Summit White  
4708                       Bright Silver

In [21]:
df.loc[[False, True, False, True, False, True, False], [False, True, True, False]] # или iloc

miles                fuel_type
4703       10 miles                   Diesel
4705  211,853 miles                 Gasoline
4707        4 miles   Compressed Natural Gas

In [22]:
df.loc[[False, True, True, True, True, False, False]] # или iloc

price          miles  fuel_type                      exterior_color
4703  $54,647       10 miles     Diesel   Brilliant Black Crystal Pearlcoat
4704   $8,488  213,601 miles    Unknown                               White
4705   $2,500  211,853 miles   Gasoline                                Blue
4706   $9,950    6,165 miles     Diesel                                 Red

In [23]:
df.loc[df.exterior_color == ' Red',  ]

price        miles fuel_type exterior_color
4706  $9,950  6,165 miles    Diesel            Red

**Обзор данных:**
    
С помощью метода Dataframe.describe() библиотеки pandas, осмотрим наши столбцы данных на наличе пустых столбов, аномалий или закономерностей

**Для номинальных признаков**
count - количество значений

unique - количество уникальных значений в каждом столбце

top - самое распространенное значение

freq - частота наиболее распространенного значения

**Для количественных признаков:**
mean - Среднее арифметическое

std - Стандартное отклонение

min - Минимальное значение

25% - Квартиль уровня (1/4)

50% - Медиана

75% - Квартиль уровня (3/4)

max - Максимальное значение

In [24]:
full_data.iloc[:,:13].describe(include="all")

price    miles  fuel_type exterior_color   interior_color  \
count     50987    42301      51655          50885            45102   
unique    23114    18609          7           1124             1268   
top     $13,995  5 miles   Gasoline   Shadow Black   Charcoal Black   
freq        107     3582      38282           5387             8720   
mean        NaN      NaN        NaN            NaN              NaN   
std         NaN      NaN        NaN            NaN              NaN   
min         NaN      NaN        NaN            NaN              NaN   
25%         NaN      NaN        NaN            NaN              NaN   
50%         NaN      NaN        NaN            NaN              NaN   
75%         NaN      NaN        NaN            NaN              NaN   
max         NaN      NaN        NaN            NaN              NaN   

       drivetrain        transmission                  engine  \
count       48653               49856                   49340   
unique         66                 343                    1493   
top           4WD   6-Speed Automatic   3.5L V6 24V MPFI DOHC   
freq        17845               20154                    2863   
mean          NaN                 NaN                     NaN   
std           NaN                 NaN                     NaN   
min           NaN                 NaN                     NaN   
25%           NaN                 NaN                     NaN   
50%           NaN                 NaN                     NaN   
75%           NaN                 NaN                     NaN   
max           NaN                 NaN                     NaN   

                       VIN                     name  \
count                51526                    51657   
unique               50839                     3299   
top      NM0AE8F70F1193310  New 2018 Ford F-150 XLT   
freq                     8                     2080   
mean                   NaN                      NaN   
std                    NaN                      NaN   
min                    NaN                      NaN   
25%                    NaN                      NaN   
50%                    NaN                      NaN   
75%                    NaN                      NaN   
max                    NaN                      NaN   

                                sellerAddress            id modelName  
count                                   51517  5.178300e+04     51783  
unique                                   2468           NaN        41  
top     900 Bernadette Dr, Columbia, MO 65203           NaN     F-150  
freq                                      708           NaN      4742  
mean                                      NaN  7.472646e+08       NaN  
std                                       NaN  1.459627e+07       NaN  
min                                       NaN  2.648982e+06       NaN  
25%                                       NaN  7.447334e+08       NaN  
50%                                       NaN  7.508864e+08       NaN  
75%                                       NaN  7.540559e+08       NaN  
max                                       NaN  7.564938e+08       NaN

Самое встречаемое значение наших классов состояния машин - 4WD, количество непропущенных строк даных - 48653, а частота входжения - 17845, что составляет - 34% значений всего класса, а значит наши классы по этому столбцу - сбалансированны (при условии что баланс считаем нарушенным если, один класс занимает более 75% от класса)

**Визуализация данных и описательная статистика**

Рассмотрим диаграмму по двум количественным признакам - цене от пробегу автомобиля, предполагаем что эти данные зависимы т.к с ростом пробега, будет уменьшатся цена автомобиля